core/src/main/java/org/lflang/federated/generator/FederateInstance.java

```diff
@@ -698,24 +719,64 @@ public TimeValue findNearestPhysicalActionTrigger(ReactionInstance reaction) {
           // Logical action
           // Follow it upstream inside the reactor
           for (ReactionInstance uReaction : actionInstance.getDependsOnReactions()) {
-            // Avoid a loop
-            if (!Objects.equal(uReaction, reaction)) {
+            // Avoid a potentially deep loop by checking the visited set.
+            if (!visited.contains(uReaction)) {
               TimeValue uMinDelay =
-                  actionInstance.getMinDelay().add(findNearestPhysicalActionTrigger(uReaction));
+                  actionInstance
+                      .getMinDelay()
+                      .add(findNearestPhysicalActionTriggerRecursive(visited, uReaction));
               if (uMinDelay.isEarlierThan(minDelay)) {
                 minDelay = uMinDelay;
               }
             }
           }
         }
-
-      } else if (trigger.getDefinition() instanceof Output) {
-        // Outputs of contained reactions
-        PortInstance outputInstance = (PortInstance) trigger;
-        for (ReactionInstance uReaction : outputInstance.getDependsOnReactions()) {
-          TimeValue uMinDelay = findNearestPhysicalActionTrigger(uReaction);
-          if (uMinDelay.isEarlierThan(minDelay)) {
-            minDelay = uMinDelay;
+      } else if (trigger.getDefinition() instanceof Port) {
+        PortInstance port = (PortInstance) trigger;
+        // Regardless of whether the port is an output or input port of
+        // a contained reactor, recurse on reactions that write to it.
+        for (ReactionInstance uReaction : port.getDependsOnReactions()) {
+          // Ports can form a loop also.
+          if (!visited.contains(uReaction)) {
+            TimeValue uMinDelay = findNearestPhysicalActionTriggerRecursive(visited, uReaction);
+            if (uMinDelay.isEarlierThan(minDelay)) {
+              minDelay = uMinDelay;
+            }
+          }
+        }
+        // The trigger is an input of a contained reactor. If it
+        // another upstream contained reactor connects to it, trace to
+        // the upstream contained reactor's output ports.
+        if (trigger.getDefinition() instanceof Input) {
```

Comment:
Is there a cleaner way to look at an input port (`PortInstance`) and get a connection delay from an upstream connection into the input port? @edwardalee @erlingrj